### ISSUE TITLE
Target suggestions only apply to first page of buffer history

### ISSUE THREAD

**Issue by nnarayen** [IngoMeyer441/tmux-easy-motion#7]
Hey! First off thanks for making this awesome plugin - I've been a user of vim easy motion for years and love to see functionality like this replicated in tmux.

I'm currently running tmux 3.1c on macOS. The motion targets work great when I'm either in normal mode or I've entered copy mode but haven't scrolled the screen. However, I've noticed that when I scroll to previous pages, the target suggestions appear on the _first_ page of the buffer and then is applied to the current page of the buffer. I've captured this behavior in a video if that helps:

https://user-images.githubusercontent.com/1946683/105400758-5ec3c180-5bf3-11eb-8b99-b3dc26dc406d.mov

Is it expected behavior that the jump targets should only work on the first page of the buffer? Or is it possible that there's something wrong with my settings?

Separately, I noticed that on macOS 10.15 and bash `GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin19)`, very few key bindings were getting set up by default. I tracked the error on my machine down to https://github.com/IngoMeyer441/tmux-easy-motion/blob/master/easy_motion.tmux#L34, where I had to replace `-N1` with `-n1` for bash compatibility. I'm not an expert by any means in shell scripting, so didn't want to put up a PR without understanding the implications first.

Thanks again for making the plugin!

**Comment by IngoMeyer441** [comment]
Hey, thanks for giving my plugin a try!
I develop on Linux but tried not to use any language features of Bash >= 4 (since macOS ships with Bash 3 for license reasons). I didn't know that `read -N` needs Bash 4 or newer. As you already suggested, `-N` can be replaced with `-n` in this case. Both only differ in handling of delimiters (which is whitespace by default) and does not matter for reading a single character. This is fixed in commit e6f06655ac32a0ca5926c7dbdc288eec2c5ad5b4.
Thanks for pointing me to the bug of a wrong scroll position, I simply didn't notice this yet. This is fixed in commit c69c39d7fa9fb6ace89cfb6bd4ef169cb94d463b. Both commits are already merged into the `master` branch, so you can update the plugin and (hopefully :wink:) everything works as expected now.

**Comment by nnarayen** [comment]
Hey thanks for the quick response! Sorry about the delay on my end, haven't had a chance to test this out.

I pulled the most recent changes (on commit `a3a80340e00c8387f19b053620cf95ec9a2d4f5f`), and while they seem have to have fixed the suggestions being applied to the wrong screen of the scroll buffer, I'm still experiencing some weird issues.

The following screen recording demonstrates both:

- All movement seems to think that the cursor is on the _lowest_ row in the buffer. In the screen recording, in both examples I'm attempting to jump to **row 2** of the file in the buffer history and therefore hit target `;v`. The first time when I hit the easy motion prefix from normal mode, the cursor is on the lowest row and it works as expected. However, when I enter copy mode and move the cursor one row up and hit the same target keystrokes, it actually jumps me to row one. If I move up to the second row from the bottom, my final position will also be two positions above the target I hit. It feels like part of the same bug that the target suggestions stay the same even when I move the cursor one row up, which theoretically changes the number of rows before my cursor. 
- The only commands that seem to work (in both normal and copy mode) are the ones that highlight targets _before_ my cursor location. These commands are easy-motion-prefix + B or easy-motion-prefix + b. It also highlights words across the entire buffer screen even when my cursor is in the middle.

https://user-images.githubusercontent.com/1946683/105938052-cc715280-6024-11eb-81f6-d3d99cb6cf26.mov

Please let me know if there is any additional information I can provide to help with debugging / reproducibility!

**Comment by IngoMeyer441** [comment]
Ok, that is really strange. On my system I cannot reproduce your issues :disappointed:. Things that are different on our machines:
- You use macOS, I use (Manjaro) Linux
- You use Alacritty as a terminal emulator, I use kitty
Do you have any other special setup?

**Comment by IngoMeyer441** [comment]
Ok, I have an idea. Tmux uses emacs style key bindings by default but for this plugin you must have activated the vi key bindings. Do you have the setting
```
set -g mode-keys vi
```
in your `.tmux.conf`?

**Comment by nnarayen** [comment]
Yeah I've been wondering if it's a macOS issue, I wish I knew more about tmux to play around with the plugin to debug. For what it's worth, I can reproduce this behavior on my system with iTerm2 3.3.12. 

You're right, I do have vi style key bindings set here https://github.com/nnarayen/dotfiles/blob/master/tmux/tmux.conf#L88

**Comment by IngoMeyer441** [comment]
I have run a test on a fresh macOS catalina + Homebrew + tmux + tpm and it worked for me with a minimal configuration (loading tpm, settinging vi mode keys and loading tmux-easy-motion).

**Comment by IngoMeyer441** [comment]
I have tested in the stock Apple terminal emulator.

**Comment by nnarayen** [comment]
Ah perfect! I'll try removing things from my tmux config until I can reproduce the correct behavior. Right now, I seem to have a weird keybinding from one of my plugins:

```
bind-key    -T copy-mode          Space                send-keys -X page-down
```

I don't use the default copy mode, but the fact that my cursor keeps getting sent to the bottom of the screen seems related. I'll report back here with what I find! Thanks for all the help investigating 😄

**Comment by nnarayen** [comment]
Well this is going to be really unsatisfying, I started with a minimal conf that just enabled tpm + navigator + copy mode to confirm the correct behavior on my system. After that, I slowly brought back different parts of my configuration but I could _never_ reproduce my old behavior again..

So now I have exactly the same config as [earlier](https://github.com/nnarayen/dotfiles/blob/master/tmux/tmux.conf) but everything works as expected. The only thing I can think of that might have had an effect was I deleted the extracto plugin since its leader key was [conflicting](https://github.com/nnarayen/dotfiles/blob/master/tmux/tmux.conf#L117) with easy motion, but even after restoring the conflict I see the correct behavior. I got my laptop within the last two weeks, so it doesn't seem possible that [upgrading extracto](https://github.com/laktak/extrakto/commits/master) would have fixed the issue either.

I wish I could offer more insight into what was going on, but it definitely seems like a weird issue on my end! Thanks again for creating a great plugin and helping me debug my issues 😄
